### PR TITLE
Feature: xray watch

### LINKED ISSUE #202: Feature: Support xRay Watches
Hi,

I'd like to be able to configure xRay watches using this library.
This would allow me to consume it from the artifactory terraform provider.

If I were to implement this, should it be created in `./artifactory/services/xraywatches.go` ?

## References:
- https://www.jfrog.com/confluence/display/JFROG/Xray+REST+API#XrayRESTAPI-WATCHES
- https://github.com/xero-oss/go-xray/blob/master/xray/v2/watches.go

### PR DESCRIPTION
**Pull request by josh-barker-coles**
## Description:

This PR creates:
- an Xray Manager that uses a separate endpoint
- a function to retrieve the Xray Version - `GetXrayVersion`
- functions to Create, Get, Update and Delete an Xray Watch
  - `CreateXrayWatch`
  - `GetXrayWatch`
  - `UpdateXrayWatch`
  - `DeleteXrayWatch`

The Xray Watch functions support the configuration of Repositories and Builds.

There is a placeholder method to configure Bundles.
I can't implement the support for Bundles because of the limits on the free trial for artifactory.

## Related Issues

Resolves #202

### CODE REVIEW

**Comment by josh-barker-coles** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by josh-barker-coles** [comment]
recheckcla

**Comment by josh-barker-coles** [comment]
recheckcla

**Comment by josh-barker-coles** [comment]
Hi @eyalbe4, it would be good to get your review on this.

Cheers

**Comment by eyalbe4** [comment]
Sure @josh-barker-coles.
I'll get to reviewing this soon.

File: xray/services/utils/xraywatchbody.go [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+	case "":
+		// empty is fine
+	default:
+		return errors.New("Invalid Build Type. Must be " + string(WatchBuildAll) + " or " + string(WatchBuildByName))
+	}
+
+	return nil
+}
+
+func configureBundles(payloadBody *XrayWatchBody, params XrayWatchParams) error {
```
Comment (eyalbe4) [inline xray/services/utils/xraywatchbody.go:308]: We prefer not to have placeholder code in the form of empty methods. I suggest you to the description of the ```CreateBody``` function a comment saying that bundles are currently not supported. This disclaimer should also be added to the API description in the README.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
I haven't gone through everything yet. Releasing what I have so far.

File: tests/xraywatch_test.go [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+	return nil
+}
+
+func deletePolicy(policyName string) error {
+	xrayDetails := GetXrayDetails()
+	xrayHTTPDetails := xrayDetails.CreateHttpClientDetails()
+	artUtils.SetContentType("application/json", &xrayHTTPDetails.Headers)
+	client, err := httpclient.ClientBuilder().Build()
+	if err != nil {
+		return nil
```
Comment (eyalbe4) [inline tests/xraywatch_test.go:543]: Should probably be fixed to ```return err```.

File: tests/xraywatch_test.go [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+	}
+
+	resp, _, err := client.SendDelete(xrayDetails.GetUrl()+"api/v2/policies/"+policyName, nil, xrayHTTPDetails)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		return errors.New("failed to delete policy " + resp.Status)
+	}
```
Comment (eyalbe4) [inline tests/xraywatch_test.go:549]: I suggest changing the above to:
```
	if err != nil  {
		return error
	}
	if resp.StatusCode != http.StatusOK {
		return errors.New("failed to delete policy " + resp.Status)
	}
```
This also applies to other checks in this file.

File: xray/services/xraywatch.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
+package services
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	rthttpclient "github.com/jfrog/jfrog-client-go/artifactory/httpclient"
+	// "github.com/jfrog/jfrog-client-go/artifactory/services/utils"
```
Comment (eyalbe4) [inline xray/services/xraywatch.go:9]: Let's remove this commented line.

File: xray/services/xraywatch.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"github.com/jfrog/jfrog-client-go/xray/services/utils"
+)
+
+const (
+	watchAPIURL = "api/v2/watches"
+)
+
+// XrayWatchService defines the http client and xray details
+type XrayWatchService struct {
```
Comment (eyalbe4) [inline xray/services/xraywatch.go:24]: Ket's rename the struct to ```WatchService``` and this file to ```watch.go```.
Let's also rename ```NewXrayWatchService``` to ```NewWatchService```.

File: xray/services/xraywatch.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+func NewXrayWatchService(client *rthttpclient.ArtifactoryHttpClient) *XrayWatchService {
+	return &XrayWatchService{client: client}
+}
+
+// GetJfrogHttpClient returns the http client
+func (xws *XrayWatchService) GetJfrogHttpClient() *rthttpclient.ArtifactoryHttpClient {
+	return xws.client
+}
+
+func (xws *XrayWatchService) getXrayWatchURL() string {
```
Comment (eyalbe4) [inline xray/services/xraywatch.go:39]: Let's rename ```getXrayWatchURL()``` to ```getWatchURL()```.

File: xray/services/xraywatch.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	log.Info("Deleting watch...")
+	resp, body, err := xws.client.SendDelete(xws.getXrayWatchURL()+"/"+watchName, nil, &httpClientsDetails)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return errorutils.CheckError(errors.New("Artifactory response: " + resp.Status + "\n" + clientutils.IndentJson(body)))
+	}
+
+	log.Debug("Artifactory response:", resp.Status)
```
Comment (eyalbe4) [inline xray/services/xraywatch.go:56]: Ket's change all ```Artifactory response``` messages in this file to ```Xray response```.

File: xray/services/xraywatch.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+	log.Info("Creating watch...")
+	resp, respBody, err = xws.client.SendPost(url, content, &httpClientsDetails)
+
+	log.Info("Finished request")
+	if err != nil {
+		log.Info("err: " + err.Error())
+		log.Error("error")
+		return err
+	}
+	log.Info("statuscode: " + resp.Status)
```
Comment (eyalbe4) [inline xray/services/xraywatch.go:88]: Let's remove the above logging lines, and just return the error if not nil.

File: xray/services/xraywatch.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+// Update will update an existing Xray watch by name
+// It will error if no watch can be found by that name.
+func (xws *XrayWatchService) Update(params utils.XrayWatchParams) error {
+	payloadBody, err := utils.CreateBody(params)
+	if err != nil {
+		return errorutils.CheckError(err)
+	}
+
+	// the update payload must not have a name
+	payloadBody.GeneralData.Name = ""
```
Comment (eyalbe4) [inline xray/services/xraywatch.go:106]: Why is this needed?

File: xray/services/xraywatch.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	content, err := json.Marshal(payloadBody)
+	if err != nil {
+		return errorutils.CheckError(err)
+	}
+
+	httpClientsDetails := xws.XrayDetails.CreateHttpClientDetails()
+	artUtils.SetContentType("application/json", &httpClientsDetails.Headers)
+	var url = xws.getXrayWatchURL() + "/" + params.Name
```
Comment (eyalbe4) [inline xray/services/xraywatch.go:119]: ```getXrayWatchURL()``` may end with a slash. Let's use ```AddTrailingSlashIfNeeded```.

File: xray/services/xraywatch.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+	log.Info("statuscode: " + resp.Status)
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
+		return errorutils.CheckError(errors.New("Artifactory response: " + resp.Status + "\n" + clientutils.IndentJson(respBody)))
+	}
+	log.Debug("Artifactory response:", resp.Status)
+	log.Info("Done updating watch.")
+	return nil
+}
+
+// Get retrieves the details about an Xray watch by name
```
Comment (eyalbe4) [inline xray/services/xraywatch.go:141]: // Get retrieves the details about an Xray watch by name --> // Get retrieves the details of an Xray watch by its name.

File: xray/manager.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+	return manager, err
+}
+
+// Client will return the http client
+func (sm *XrayServicesManager) Client() *rthttpclient.ArtifactoryHttpClient {
+	return sm.client
+}
+
+// GetXrayVersion will return the xray version
+func (sm *XrayServicesManager) GetXrayVersion() (string, error) {
```
Comment (eyalbe4) [inline xray/manager.go:41]: Let's rename ```GetXrayVersion()``` to ```GetVersion()```,  ```CreateXrayWatch``` to ```CreateWatch``` and so on.

File: README.md [hunk 13]
```diff
@@ -89,6 +89,16 @@
...
+- [Xray APIs](#xray-apis)
+  - [Creating Xray Service Manager](#creating-xray-service-manager)
+    - [Creating Xray Details](#creating-xray-details)
+    - [Creating Xray Service Config](#creating-xray-service-config)
+    - [Creating New Xray Service Manager](#creating-new-xray-service-manager)
+  - [Using Xray Services](#using-xray-services)
+    - [Creating an Xray Watch](#creating-an-xray-watch)
+    - [Get an Xray Watch](#get-an-xray-watch)
+    - [Update an Xray Watch](#update-an-xray-watch)
+    - [Delete an Xray Watch](#delete-an-xray-watch)
```
Comment (eyalbe4) [inline README.md:101]: Let's also include the new version API in the README.

File: README.md [hunk 14]
```diff
@@ -1058,4 +1069,83 @@ resultReader.Reset()
...
+xrayDetails.SetAccessToken("accesstoken")
+// if client certificates are required
+xrayDetails.SetClientCertPath("path/to/.cer")
+xrayDetails.SetClientCertKeyPath("path/to/.key")
+```
+
+#### Creating Xray Service Config
+```go
+serviceConfig, err := config.NewConfigBuilder().
+    SetServiceDetails(rtDetails).
```
Comment (eyalbe4) [inline README.md:1093]: rtDetails --> xrayDetails

File: README.md [hunk 15]
```diff
@@ -1058,4 +1069,83 @@ resultReader.Reset()
...
+    SetCertificatesPath(certPath).
+    SetThreads(threads).
+    SetDryRun(false).
```
Comment (eyalbe4) [inline README.md:1096]: Let'e remove ```SetThreads``` and```SetDryRun```

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
Done reviewing everything.
Thanks so much for adding this!
See my inline comments.

**Comment by josh-barker-coles** [comment]
Hey @eyalbe4, thanks for the feedback. I'll start working on it this week.

File: xray/services/utils/xraywatchbody.go [hunk 16]
```diff
@@ -0,0 +1,429 @@
...
+}
+
+// XrayWatchBody is the top level payload to be sent to xray
+type XrayWatchBody struct {
+	GeneralData      xrayWatchGeneralParams    `json:"general_data"`
+	ProjectResources xrayWatchProjectResources `json:"project_resources,omitempty"`
+	AssignedPolicies []XrayAssignedPolicy      `json:"assigned_policies,omitempty"`
+}
+
+// structs that aren't exported
```
[the remark this answers is not in the document]
Comment (josh-barker-coles) [reply xray/services/utils/xraywatchbody.go:101]: @eyalbe4 
If a struct or function starts with an upper case, it's available in the package to be imported somewhere else.
However, if a struct or function starts with a lower case, it's internal and cannot be imported elsewhere.

As these structs are used for internally transforming the parameters to the payload, I opted to not export them.
What would you prefer?

File: xray/services/xraywatch.go (hunk 9 above)
Comment (josh-barker-coles) [reply xray/services/xraywatch.go:106]: @eyalbe4 

Xray responds with an error when the watch's name is set and you attempt to update the watch.

The example payload does not have the name field set under `general_data` https://www.jfrog.com/confluence/display/XRAY2X/Xray+REST+API#XrayRESTAPI-UpdateWatch

File: xray/services/xraywatch.go (hunk 5 above)
Comment (josh-barker-coles) [reply xray/services/xraywatch.go:24]: @eyalbe4 

Most, if not all, structs have a prefix of `Xray`.
Did you want them all to be renamed without Xray?

File: tests/utils.go [hunk 17]
```diff
@@ -158,6 +167,14 @@ func createDistributionManager() {
 	testsBundleDeleteRemoteService.DistDetails = distDetails
 }
 
+func createXrayManager() {
+	xrayDetails := GetXrayDetails()
+	client, err := rthttpclient.ArtifactoryClientBuilder().SetServiceDetails(&xrayDetails).Build()
```
[the remark this answers is not in the document]
Comment (josh-barker-coles) [reply tests/utils.go:172]: Hey @eyalbe4,

I understand why you want to move that functionality, but I've noticed a few things.

- In the root directory, there already is a generic [httpclient](https://github.com/jfrog/jfrog-client-go/tree/master/httpclient) package, which [ArtifactoryHttpClient](https://github.com/jfrog/jfrog-client-go/blob/master/artifactory/httpclient/httpclient.go#L15) depends on.
- The Distribution manager is also depend on the [ArtifactoryClientBuilder](https://github.com/jfrog/jfrog-client-go/blob/master/distribution/manager.go#L4)

This sounds like a big change that should go in a separate PR, as it would effect many files under `artifactory/*`, `distribution/*`, `tests/*` and `xray/*`.

As the `httpclient` directory & package is already taken, what did you want to do?

Cheers

**Comment by josh-barker-coles** [comment]
Hi @eyalbe4, I'll fix the README conflict after we've resolved all the above discussions.

File: xray/services/xraywatch.go (hunk 5 above)
Comment (josh-barker-coles) [reply xray/services/xraywatch.go]: Hi @eyalbe4, I've renamed all structs & functions that included Xray.

File: tests/utils.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (eyalbe4) [reply tests/utils.go:172]: Got it @josh-barker-coles. We'll create an issue for this and handle this in a follow up PR.

**Comment by josh-barker-coles** [comment]
Hi @eyalbe4, what else is remaining for this PR?
I'd like to complete this so I can work with someone on https://github.com/jfrog/terraform-provider-artifactory/pull/10

Cheers,

Josh

**Review by eyalbe4 — APPROVED** [review]
Thanks so much for this awesome contribution @josh-barker-coles!
Please go ahead and resolve the conflict, and I'll merge this PR.

**Comment by josh-barker-coles** [comment]
You're welcome @eyalbe4.

I've just rebased & resolved the README.md issue.

**Comment by yahavi** [comment]
@josh-barker-coles 
FYI, the APIs for Xray Watches in the Go Client are going to change in the next version - We are removing the HTTP response from the returned values. 
We think that returning the HTTP response is "too low level" for such APIs. If the API failed to create the watch, an error is returned.

Before the change:
```go
func (xws *WatchService) Delete(watchName string) (*http.Response, error)
func (xws *WatchService) Create(params utils.WatchParams) (*http.Response, error)
func (xws *WatchService) Update(params utils.WatchParams) (*http.Response, error)
func (xws *WatchService) Get(watchName string) (watchResp *utils.WatchParams, resp *http.Response, err error)
```

After that changes:
```go
func (xws *WatchService) Delete(watchName string) error
func (xws *WatchService) Create(params utils.WatchParams) error
func (xws *WatchService) Update(params utils.WatchParams) error
func (xws *WatchService) Get(watchName string) (watchResp *utils.WatchParams, err error)
```

We hope for your understanding.

**Comment by josh-barker-coles** [comment]
Hi @yahavi

Thanks for letting me know.
However, I think its worth discussing the rational behind returning the HTTP response.

Although an error indicates that something has gone wrong, it limits the ability of consumers to handle error conditions.
For example, a Internal server error (500), a Not Found (404) error, and an Unauthorized (401) error are all quite different and may require different actions.

The implication of removing the HTTP response will force consumers to write their own code to parse the string and attempt to handle the generic error.

Additionally, this design decision has wider implications for the artifactory terraform provider https://github.com/jfrog/terraform-provider-artifactory as the future goal is to refactor it to use this client library.
Yet, the 2 clients have different designs in that the old client (www.github.com/atlassian/go-artifactory) returns HTTP Responses, yet the new (jfrog-client-go) does not.

One example is https://github.com/jfrog/terraform-provider-artifactory/blob/master/pkg/artifactory/resource_artifactory_virtual_repository.go#L151-L157


Looking forward to your responses

/cc: @eyalbe4  @DarthFennec @chb0github

**Comment by eyalbe4** [comment]
Thank you for sharing this input @josh-barker-coles.
We may be able to support this requirement by attaching the HTTP response to the error returned by the APIs. We can build this mechanism in a way that it will seamlessly serve all APIs, existing and future ones.
This apparoach may require a some more thought, but I think this is doable.
Please let me know what you think.

**Comment by josh-barker-coles** [comment]
Hi @eyalbe4, no worries.
Yep, I think that would be great. Do you have an implementation in mind?

It might be good to discuss with the maintainers of https://github.com/jfrog/terraform-provider-artifactory to see how that could work.

Cheers
